# 🧠 Lab7 - IR Basics & Vector Space Proximity

**Team:** Group 6  
**Members:** Emmanuel Ihejiamaizu · Liggia Elena Taboada Cruz · Chao-Chung Liu (Thomas)  

---

## 🔹 Student Tasks
### [Part A — Build the Corpus](#part_a)
Download a large corpus and create a document collection.

You must report:

- corpus source
- number of documents
- approximate vocabulary size
- a short description of the domain

### [Part B — Build the Retrieval Pipeline](#part_b)

Implement the following:

1. **[Tokenizer](#tokenizer)** 
2. **[Normalization](#normalization)**
3. **[Stop-word removal](#stopwordremoval)**
4. **[Stemming or lemmatization](#stemmingorlemmatization)**
5. **[Term-document incidence matrix](#termdocumentincidencematrix)**
6. **[Term frequency](#termfrequency)**
7. **[Log-frequency weighting](#logfrequencyweighting)**
8. **[Document frequency](#documentfrequency)**
9. **[Inverse document frequency](#inversedocumentfrequency)**
10. **[TF-IDF weighting](#tfidfweighting)**
11. **[Cosine similarity retrieval](#cosinesimilarityretrieval)**


### [Part C — Querying](#part_c)

Create at least **[5 information needs](#5_informationneeds)** and convert each into one or more queries.

**[query_1](#query_1)**,**[query_2](#query_2)**,**[query_3](#query_3)**,**[query_4](#query_4)**,**[query_5](#query_5)**

For each query:

- retrieve the top documents
- explain why the retrieved documents are or are not relevant
- compare retrieval using at least **two different representations**, such as:
  - binary incidence
  - raw TF
  - TF-IDF


### [Part D — Evaluation](#part_d)

For at least **3 queries**, create relevance judgments and compute:

**[query_1](#query_1e)**,**[query_2](#query_2e)**,**[query_3](#query_3e)**

- confusion matrix
- precision
- recall
- F1-score
- Precision@K
- Average Precision (AP)
- Mean Reciprocal Rank (MRR)

If two teammates judge the same query-document pairs, also compute **Cohen's Kappa**.

## 🔹 Deliverables

Your notebook must include:

### 1. Corpus Description
A markdown section describing the corpus and why you selected it.

### 2. Preprocessing Pipeline
Code and explanation for each preprocessing step.

### 3. Vector Space Construction
Show examples of:

- incidence matrix
- TF values
- DF values
- IDF values
- TF-IDF matrix

### 4. Retrieval Demonstration
Show retrieval examples for multiple queries.

### 5. Evaluation Section
Show your relevance judgments and evaluation metrics.

### 6. Reflection
Answer these questions:

1. Which representation worked best and why?
2. Did TF-IDF improve over raw term counts?
3. What kinds of false positives did you observe?
4. What kinds of relevant documents were missed?
5. How did the evaluation metrics help you understand system quality?
6. How would you improve the system next?

### Team Requirement

Work in teams of 3.  
One team member must submit the final notebook and repository link before class ends.


## 🔹 Optional Extension Ideas

If your team finishes early, extend your system by trying one or more of the following:

- compare stemming vs. lemmatization
- compare different stop-word lists
- compare cosine similarity with another similarity measure
- add relevance feedback
- build a small interactive search interface
- compare performance across document categories
- create a precision–recall curve for multiple queries

### Success Criteria

A strong submission will:

- clearly connect every implementation step to the workshop concepts
- use a genuinely large corpus
- include careful interpretation of results
- explain retrieval errors, not just metric values
- demonstrate that evaluation is tied to the **information need**, not just term overlap


## 🧩 Workshop Structure (90 Minutes)

1. **Instructor-led walkthrough of Vector Space Proximity** *(20 min)*  
   - Form teams of 3  
   - Review the notebook structure and expectations  
   - Understand how TF, IDF, TF-IDF, and cosine similarity connect  
   - Ask questions before starting implementation  

2. **Team Jupyter Notebook Development** *(65 min)*  
   - Build a complete **IR pipeline**:
     - Tokenization, normalization, stop-word removal, stemming  
     - Term-Document Incidence Matrix  
     - TF, Log TF, DF, IDF, TF-IDF  
     - Cosine similarity retrieval  
   - Implement **evaluation metrics**:
     - Confusion matrix, Precision, Recall, F1  
     - Precision@K, AP, MRR  
     - (Optional) Kappa for inter-judge agreement  
   - Document all steps using Markdown  

3. **Push to GitHub** *(5 min)*  
   - One team member pushes the final notebook  
   - ⚠️ Include all team member names clearly in the notebook  

4. **Instructor Review (Live Coaching)**  
   - Instructor circulates during development  
   - Provides feedback during a **peer-review-style walkthrough**  

5. **Email Submission** *(1 min)*  
   - One email per team  
   - Include **GitHub repo link (*.git)**  
   - Subject:  
     **PROG8245 - Vector Space Proximity Workshop, Team #_____**


## 💻 Submission Checklist

- ✅ `VectorSpaceProximityWorkshop.ipynb` with:

  - Demo code:
    - Large **Document Collection**
    - **Tokenizer + Normalization Pipeline**
    - Full **Vector Space Model implementation**

  - Implementations:
    - Term-Document Incidence Matrix  
    - TF and Log Frequency Weight  
    - DF and IDF  
    - TF-IDF  
    - Cosine Similarity Retrieval  

  - Evaluation:
    - Confusion Matrix  
    - Precision, Recall, F1  
    - Precision@K  
    - Average Precision (AP)  
    - Mean Reciprocal Rank (MRR)  
    - (Optional) Kappa  

  - Markdown explanations for each major step  

  - **Labeled talking points**:
    - Explain how TF-IDF + cosine similarity improve retrieval  
    - Compare at least two representations (e.g., Binary vs TF-IDF)  

---

- ✅ `README.md` with:

  - Dataset description  
  - Team member names  
  - Link to dataset (and license if applicable)  
  - Summary of retrieval approach  

---

- ✅ GitHub Repository:

  - Public repo named:  
    `VectorSpaceProximityWorkshop`  

  - One team member publishes the repo  

  - At least **one meaningful commit** that includes:
    - a working pipeline  
    - and one clearly written technical insight  

---

### 🧠 Key Requirement

Your system must demonstrate that:

> **Relevance is evaluated with respect to the information need, not just keyword matching.**

---


<a id="part_a"></a>
## Part A — Build the Corpus

**Corpus Source:** https://www.cisa.gov/known-exploited-vulnerabilities-catalog  
**Number of Documents:** 1551  
**Approximate Vocabulary Size:** 3449  
**Domain:** Cybersecurity. The dataset contains official records of Common Vulnerabilities and Exposures (CVEs), including vendor names, affected products, vulnerability descriptions, and remediation actions.  
**Corpus Name:** known_exploited_vulnerabilities.csv  

In [70]:
import pandas as pd
import re

df = pd.read_csv('known_exploited_vulnerabilities.csv')

df.head(5)

,cveID,vendorProject,product,vulnerabilityName,dateAdded,shortDescription,requiredAction,dueDate,knownRansomwareCampaignUse,notes,cwes
0,CVE-2025-32432,Craft CMS,Craft CMS,Craft CMS Code Injection Vulnerability,2026-03-20,Craft CMS contains a code injection vulnerability that allows a remote attacker to execute arbitrary code.,"Apply mitigations per vendor instructions, follow applicable BOD 22-01 guidance for cloud services, or discontinue use of the product if mitigations are unavailable.",2026-04-03,Unknown,https://craftcms.com/knowledge-base/craft-cms-cve-2025-32432 ; https://github.com/craftcms/cms/security/advisories/GHSA-f3gw-9ww9-jmc3 ; https://nvd.nist.gov/vuln/detail/CVE-2025-32432,CWE-94
1,CVE-2025-54068,Laravel,Livewire,Laravel Livewire Code Injection Vulnerability,2026-03-20,Laravel Livewire contain a code injection vulnerability that could allow unauthenticated attackers to achieve remote command execution in specific scenarios.,"Apply mitigations per vendor instructions, follow applicable BOD 22-01 guidance for cloud services, or discontinue use of the product if mitigations are unavailable.",2026-04-03,Unknown,https://github.com/livewire/livewire/security/advisories/GHSA-29cq-5w36-x7w3 ; https://github.com/livewire/livewire/commit/ef04be759da41b14d2d129e670533180a44987dc ; https://nvd.nist.gov/vuln/detail/CVE-2025-54068,CWE-94
2,CVE-2025-43510,Apple,Multiple Products,Apple Multiple Products Improper Locking Vulnerability,2026-03-20,"Apple watchOS, iOS, iPadOS, macOS, visionOS, and tvOS contain an improper locking vulnerability that could allow a malicious application to cause unexpected changes in memory shared between processes.","Apply mitigations per vendor instructions, follow applicable BOD 22-01 guidance for cloud services, or discontinue use of the product if mitigations are unavailable.",2026-04-03,Unknown,https://support.apple.com/en-us/125632 ; https://support.apple.com/en-us/125633 ; https://support.apple.com/en-us/125634 ; https://support.apple.com/en-us/125635 ; https://support.apple.com/en-us/125636 ; https://support.apple.com/en-us/125637 ; https://support.apple.com/en-us/125638 ; https://support.apple.com/en-us/125639 ; https://nvd.nist.gov/vuln/detail/CVE-2025-43510,CWE-667
3,CVE-2025-43520,Apple,Multiple Products,Apple Multiple Products Classic Buffer Overflow Vulnerability,2026-03-20,"Apple watchOS, iOS, iPadOS, macOS, visionOS, tvOS, and iPadOS contain a classic buffer overflow vulnerability which could allow a malicious application to cause unexpected system termination or write kernel memory.","Apply mitigations per vendor instructions, follow applicable BOD 22-01 guidance for cloud services, or discontinue use of the product if mitigations are unavailable.",2026-04-03,Unknown,https://support.apple.com/en-us/125632 ; https://support.apple.com/en-us/125633 ; https://support.apple.com/en-us/125634 ; https://support.apple.com/en-us/125635 ; https://support.apple.com/en-us/125636 ; https://support.apple.com/en-us/125637 ; https://support.apple.com/en-us/125638 ; https://support.apple.com/en-us/125639 ; https://nvd.nist.gov/vuln/detail/CVE-2025-43520,CWE-120
4,CVE-2025-31277,Apple,Multiple Products,Apple Multiple Products Buffer Overflow Vulnerability,2026-03-20,"Apple Safari, iOS, watchOS, visionOS, iPadOS, macOS, and tvOS contain a buffer overflow vulnerability that could allow the processing of maliciously crafted web content which may lead to memory corruption.","Apply mitigations per vendor instructions, follow applicable BOD 22-01 guidance for cloud services, or discontinue use of the product if mitigations are unavailable.",2026-04-03,Unknown,https://support.apple.com/en-us/124147 ; https://support.apple.com/en-us/124149 ; https://support.apple.com/en-us/124152 ; https://support.apple.com/en-us/124153 ; https://support.apple.com/en-us/124155 ; https://nvd.nist.gov/vuln/detail/CVE-2025-31277,CWE-119


In [71]:
# Count the number of documents
document_count = len(df)

# Combine all text into one large string

# .fillna('') means "if data is missing, use an empty string"
all_text = " ".join(df['vulnerabilityName'].fillna('') + " " + df['shortDescription'].fillna(''))

# Regular Expression

# .lower() makes all letters small (Apple -> apple)
# \w+ means "find words/numbers"
tokens = re.findall(r'\w+', all_text.lower())

# Count unique words only (Vocabulary Size)
vocabulary_size = len(set(tokens))

print(f"Document Count: {document_count}")
print(f"Approximate Vocabulary Size: {vocabulary_size}")

Document Count: 1551
Approximate Vocabulary Size: 3449


<a id="part_b"></a>
## Part B — Build the Retrieval Pipeline

<a id="tokenizer"></a><a id="normalization"></a>

### Tokenizer & Normalization

Convert text into a list of words (tokens)

In [72]:
import re

def my_tokenizer(text):
    
    # Lowercase: Change "The" to "the"
    lowered_text = str(text).lower()
    
    # Remove punctuation: Remove characters like ! . , ?
    # [^a-z0-9\s] means "everything that is NOT a letter, number, or space"
    clean_text = re.sub(r'[^a-z0-9\s]', '', lowered_text)
    
    # Split: cut the sentence into single words.
    tokens = clean_text.split()
    
    return tokens

# --- Test---
example_sentence = "Warning: CVE-2024! Ransomware is attacking Windows."
print("Original :", example_sentence)
print("Tokens :", my_tokenizer(example_sentence))

Original : Warning: CVE-2024! Ransomware is attacking Windows.
Tokens : ['warning', 'cve2024', 'ransomware', 'is', 'attacking', 'windows']


<a id="stopwordremoval"></a>

### Stop-word removal

Remove words that have no special meaning.

In [73]:
import nltk
from nltk.corpus import stopwords

# Download the list nltk
nltk.download('stopwords')

# 2. Get the English stop words list (set)
stop_words = set(stopwords.words('english'))

def remove_stopwords(tokens):
    
    # Keep the word ONLY IF it is not in the stop_words list
    filtered_tokens = [word for word in tokens if word not in stop_words]
    
    return filtered_tokens

# --- Test ---
# We use the tokens from the previous step
sample_tokens = ["the", "system", "is", "under", "a", "ransomware", "attack"]
print("Before :", sample_tokens)
print("After :", remove_stopwords(sample_tokens))

Before : ['the', 'system', 'is', 'under', 'a', 'ransomware', 'attack']
After : ['system', 'ransomware', 'attack']


[nltk_data] Downloading package stopwords to C:\Users\ThomasLiu-
[nltk_data]     PC\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


<a id="stemmingorlemmatization"></a>

### Stemming or lemmatization

| Feature| Stemming | Lemmatization  |
| :--- | :--- | :--- |
Method |Heuristic rules |Dictionary lookup|
Speed |Very Fast |Slower |
Accuracy |Lower: May create non-real words (e.g., "vulner"). |Higher: Always produces real dictionary words. |
Example |"studies" $\rightarrow$ "studi"|"studies" $\rightarrow$ "study"|

In [74]:
import nltk
from nltk.stem import PorterStemmer, WordNetLemmatizer

# Download necessary data for Lemmatization
nltk.download('wordnet')

# Initialize tools
stemmer = PorterStemmer()
lemmatizer = WordNetLemmatizer()

# --- Function for Stemming ---
def apply_stemming(tokens):
    # Cut words
    return [stemmer.stem(word) for word in tokens]

# --- Function for Lemmatization ---
def apply_lemmatization(tokens):
    # Dictionary lookup
    return [lemmatizer.lemmatize(word) for word in tokens]

# --- Test and Compare ---
sample_tokens = ["exploited", "exploits", "exploiting", "vulnerabilities", "feet", "was"]

print(f"{'Original':<15} | {'Stemming':<15} | {'Lemmatization':<15}")
print("-" * 50)

for word in sample_tokens:
    s = stemmer.stem(word)
    l = lemmatizer.lemmatize(word)
    print(f"{word:<15} | {s:<15} | {l:<15}")

Original        | Stemming        | Lemmatization  
--------------------------------------------------
exploited       | exploit         | exploited      
exploits        | exploit         | exploit        
exploiting      | exploit         | exploiting     
vulnerabilities | vulner          | vulnerability  
feet            | feet            | foot           
was             | wa              | wa             


[nltk_data] Downloading package wordnet to C:\Users\ThomasLiu-
[nltk_data]     PC\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


In [75]:
# Create a final function for preprocessing
def final_preprocess(text):
    # Tokenize & Lowercase 
    tokens = my_tokenizer(text) 
    
    # Remove Stopwords 
    tokens = remove_stopwords(tokens)
    
    # Lemmatization
    # Using Lemmatizer to get the real root words
    clean_tokens = [lemmatizer.lemmatize(word) for word in tokens]
    
    return " ".join(clean_tokens)

# Apply to the 'shortDescription' column
df['shortDescription_new'] = df['shortDescription'].apply(final_preprocess)

# Show the result
df[['shortDescription', 'shortDescription_new']].head()

,shortDescription,shortDescription_new
0,Craft CMS contains a code injection vulnerability that allows a remote attacker to execute arbitrary code.,craft cm contains code injection vulnerability allows remote attacker execute arbitrary code
1,Laravel Livewire contain a code injection vulnerability that could allow unauthenticated attackers to achieve remote command execution in specific scenarios.,laravel livewire contain code injection vulnerability could allow unauthenticated attacker achieve remote command execution specific scenario
2,"Apple watchOS, iOS, iPadOS, macOS, visionOS, and tvOS contain an improper locking vulnerability that could allow a malicious application to cause unexpected changes in memory shared between processes.",apple watchos io ipados macos visionos tvos contain improper locking vulnerability could allow malicious application cause unexpected change memory shared process
3,"Apple watchOS, iOS, iPadOS, macOS, visionOS, tvOS, and iPadOS contain a classic buffer overflow vulnerability which could allow a malicious application to cause unexpected system termination or write kernel memory.",apple watchos io ipados macos visionos tvos ipados contain classic buffer overflow vulnerability could allow malicious application cause unexpected system termination write kernel memory
4,"Apple Safari, iOS, watchOS, visionOS, iPadOS, macOS, and tvOS contain a buffer overflow vulnerability that could allow the processing of maliciously crafted web content which may lead to memory corruption.",apple safari io watchos visionos ipados macos tvos contain buffer overflow vulnerability could allow processing maliciously crafted web content may lead memory corruption


<a id="termdocumentincidencematrix"></a>

### Term-document incidence matrix (0/1)

In [76]:
import pandas as pd
from sklearn.feature_extraction.text import CountVectorizer

# 10 documents to see the matrix clearly
df.head(5)

,cveID,vendorProject,product,vulnerabilityName,dateAdded,shortDescription,requiredAction,dueDate,knownRansomwareCampaignUse,notes,cwes,shortDescription_new
0,CVE-2025-32432,Craft CMS,Craft CMS,Craft CMS Code Injection Vulnerability,2026-03-20,Craft CMS contains a code injection vulnerability that allows a remote attacker to execute arbitrary code.,"Apply mitigations per vendor instructions, follow applicable BOD 22-01 guidance for cloud services, or discontinue use of the product if mitigations are unavailable.",2026-04-03,Unknown,https://craftcms.com/knowledge-base/craft-cms-cve-2025-32432 ; https://github.com/craftcms/cms/security/advisories/GHSA-f3gw-9ww9-jmc3 ; https://nvd.nist.gov/vuln/detail/CVE-2025-32432,CWE-94,craft cm contains code injection vulnerability allows remote attacker execute arbitrary code
1,CVE-2025-54068,Laravel,Livewire,Laravel Livewire Code Injection Vulnerability,2026-03-20,Laravel Livewire contain a code injection vulnerability that could allow unauthenticated attackers to achieve remote command execution in specific scenarios.,"Apply mitigations per vendor instructions, follow applicable BOD 22-01 guidance for cloud services, or discontinue use of the product if mitigations are unavailable.",2026-04-03,Unknown,https://github.com/livewire/livewire/security/advisories/GHSA-29cq-5w36-x7w3 ; https://github.com/livewire/livewire/commit/ef04be759da41b14d2d129e670533180a44987dc ; https://nvd.nist.gov/vuln/detail/CVE-2025-54068,CWE-94,laravel livewire contain code injection vulnerability could allow unauthenticated attacker achieve remote command execution specific scenario
2,CVE-2025-43510,Apple,Multiple Products,Apple Multiple Products Improper Locking Vulnerability,2026-03-20,"Apple watchOS, iOS, iPadOS, macOS, visionOS, and tvOS contain an improper locking vulnerability that could allow a malicious application to cause unexpected changes in memory shared between processes.","Apply mitigations per vendor instructions, follow applicable BOD 22-01 guidance for cloud services, or discontinue use of the product if mitigations are unavailable.",2026-04-03,Unknown,https://support.apple.com/en-us/125632 ; https://support.apple.com/en-us/125633 ; https://support.apple.com/en-us/125634 ; https://support.apple.com/en-us/125635 ; https://support.apple.com/en-us/125636 ; https://support.apple.com/en-us/125637 ; https://support.apple.com/en-us/125638 ; https://support.apple.com/en-us/125639 ; https://nvd.nist.gov/vuln/detail/CVE-2025-43510,CWE-667,apple watchos io ipados macos visionos tvos contain improper locking vulnerability could allow malicious application cause unexpected change memory shared process
3,CVE-2025-43520,Apple,Multiple Products,Apple Multiple Products Classic Buffer Overflow Vulnerability,2026-03-20,"Apple watchOS, iOS, iPadOS, macOS, visionOS, tvOS, and iPadOS contain a classic buffer overflow vulnerability which could allow a malicious application to cause unexpected system termination or write kernel memory.","Apply mitigations per vendor instructions, follow applicable BOD 22-01 guidance for cloud services, or discontinue use of the product if mitigations are unavailable.",2026-04-03,Unknown,https://support.apple.com/en-us/125632 ; https://support.apple.com/en-us/125633 ; https://support.apple.com/en-us/125634 ; https://support.apple.com/en-us/125635 ; https://support.apple.com/en-us/125636 ; https://support.apple.com/en-us/125637 ; https://support.apple.com/en-us/125638 ; https://support.apple.com/en-us/125639 ; https://nvd.nist.gov/vuln/detail/CVE-2025-43520,CWE-120,apple watchos io ipados macos visionos tvos ipados contain classic buffer overflow vulnerability could allow malicious application cause unexpected system termination write kernel memory
4,CVE-2025-31277,Apple,Multiple Products,Apple Multiple Products Buffer Overflow Vulnerability,2026-03-20,"Apple Safari, iOS, watchOS, visionOS, iPadOS, macOS, and tvOS contain a buffer overflow vulnerability that could allow the processing of ma

In [77]:
# Initialize CountVectorizer with "binary=True" (this creates 0 and 1)
vectorizer = CountVectorizer(binary=True)

# Create the matrix
incidence_matrix = vectorizer.fit_transform(df['shortDescription_new'])

# Convert to DataFrame,Index is CVE ID , Columns are Words
df_incidence = pd.DataFrame(
    incidence_matrix.toarray(), 
    columns=vectorizer.get_feature_names_out(),
    index=df['cveID']
)

# Show matrix (first 5 rows and 10 words)
print("Term-Document Incidence Matrix (0/1):")
df_incidence.iloc[:5, :10]

Term-Document Incidence Matrix (0/1):


,0540,10,100,100050,10103,10530,11,11014,11302,11306
cveID,,,,,,,,,,
CVE-2025-32432,0,0,0,0,0,0,0,0,0,0
CVE-2025-54068,0,0,0,0,0,0,0,0,0,0
CVE-2025-43510,0,0,0,0,0,0,0,0,0,0
CVE-2025-43520,0,0,0,0,0,0,0,0,0,0
CVE-2025-31277,0,0,0,0,0,0,0,0,0,0


In [78]:
english_vocab = [col for col in df_incidence.columns if any(c.isalpha() for c in col)]

print("Words containing letters:", english_vocab[:10])

df_incidence[english_vocab].iloc[:5, :10]

Words containing letters: ['16bit', '2d', '32bit', '3d', '421a4ad2615941b177b6ec4ab5e239c14e62af2ab07c6df1741e2a62223223c4', '4500x', '50w', '6900w', '7zip', '9004tcp']


,16bit,2d,32bit,3d,421a4ad2615941b177b6ec4ab5e239c14e62af2ab07c6df1741e2a62223223c4,4500x,50w,6900w,7zip,9004tcp
cveID,,,,,,,,,,
CVE-2025-32432,0,0,0,0,0,0,0,0,0,0
CVE-2025-54068,0,0,0,0,0,0,0,0,0,0
CVE-2025-43510,0,0,0,0,0,0,0,0,0,0
CVE-2025-43520,0,0,0,0,0,0,0,0,0,0
CVE-2025-31277,0,0,0,0,0,0,0,0,0,0


In [79]:
row_0 = df_incidence.iloc[0]

# Find words where value is 1
words_present = row_0[row_0 == 1].index.tolist()

print(f"CVE ID: {df_incidence.index[0]}")
print(f"Value = 1: {words_present} ")

CVE ID: CVE-2025-32432
Value = 1: ['allows', 'arbitrary', 'attacker', 'cm', 'code', 'contains', 'craft', 'execute', 'injection', 'remote', 'vulnerability'] 


<a id="termfrequency"></a>

### Term frequency (TF)

In [80]:
from sklearn.feature_extraction.text import CountVectorizer
import pandas as pd

# Use CountVectorizer (Default binary=False)
tf_vectorizer = CountVectorizer()

# Create the TF matrix
tf_matrix = tf_vectorizer.fit_transform(df['shortDescription_new'])

# Create a DataFrame to show the counts 1, 2, 3...
df_tf = pd.DataFrame(
    tf_matrix.toarray(), 
    columns=tf_vectorizer.get_feature_names_out(),
    index=df['cveID']
)

# Show the first 5 rows and 10 words
print("Term Frequency (TF) Matrix - Word Counts:")
display(df_tf.iloc[:5, :10])

Term Frequency (TF) Matrix - Word Counts:


,0540,10,100,100050,10103,10530,11,11014,11302,11306
cveID,,,,,,,,,,
CVE-2025-32432,0,0,0,0,0,0,0,0,0,0
CVE-2025-54068,0,0,0,0,0,0,0,0,0,0
CVE-2025-43510,0,0,0,0,0,0,0,0,0,0
CVE-2025-43520,0,0,0,0,0,0,0,0,0,0
CVE-2025-31277,0,0,0,0,0,0,0,0,0,0


In [81]:
# Sum the counts for each word (across all documents)
word_counts = df_tf.sum(axis=0)

# Sort the words by frequency (High to Low)
top_words = word_counts.sort_values(ascending=False)

# Show the Top 10 most frequent words

print("Top 10 Most Frequent Words:")
print("-"*50)
top_words.head(10)

Top 10 Most Frequent Words:
--------------------------------------------------


vulnerability    1601
contains          886
attacker          880
allows            766
remote            731
code              722
could             477
allow             424
execution         419
microsoft         379
dtype: int64

<a id="logfrequencyweighting"></a>

### Log-frequency weighting (Log TF)

Raw TF is not ideal — a term appearing 10 times is not necessarily 10× more important than one appearing once.

**Log TF** dampens the effect of high-frequency terms:

$$w(t, d) = \begin{cases} 1 + \log_{10}(tf(t,d)) & \text{if } tf(t,d) > 0 \\ 0 & \text{otherwise} \end{cases}$$

In [82]:
import numpy as np

# Apply the formula: 1 + log10(tf) if tf > 0, else 0
def compute_log_tf(tf):
    if tf > 0:
        # We use np.log10 for the base 10 log
        return 1 + np.log10(tf)
    else:
        return 0

# Apply the function to our TF Matrix
# .applymap() moves through every single cell in the table

#df_log_tf = df_tf.applymap(compute_log_tf) # old Pandas ver. <2.1.0
df_log_tf = df_tf.map(compute_log_tf)

# Show the difference
print("Log-frequency Weighting Matrix (Log TF):")
df_log_tf.iloc[:5, :10]

Log-frequency Weighting Matrix (Log TF):


,0540,10,100,100050,10103,10530,11,11014,11302,11306
cveID,,,,,,,,,,
CVE-2025-32432,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
CVE-2025-54068,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
CVE-2025-43510,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
CVE-2025-43520,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
CVE-2025-31277,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0


In [83]:
first_row = df_log_tf.iloc[0]
first_row[first_row > 0]

allows           1.00000
arbitrary        1.00000
attacker         1.00000
cm               1.00000
code             1.30103
contains         1.00000
craft            1.00000
execute          1.00000
injection        1.00000
remote           1.00000
vulnerability    1.00000
Name: CVE-2025-32432, dtype: float64

<a id="documentfrequency"></a>

### Document frequency (DF)

Definition:How many documents (CVE records) have this word?

In [84]:
# Check if the word exists (1) or not (0)
df_exists = (df_tf > 0)

# Sum up the rows for each column (each word)
# axis=0 means "sum vertically"
document_frequency = df_exists.sum(axis=0)

# Create a DataFrame
df_df_results = pd.DataFrame(document_frequency, columns=['DF_Count'])

# Show the Top 10 words
print("Document Frequency (DF) - Top 10:")
df_df_results.sort_values(by='DF_Count', ascending=False).head(10)

Document Frequency (DF) - Top 10:


,DF_Count
vulnerability,1336
contains,886
attacker,830
allows,756
remote,701
code,674
could,432
allow,413
execution,409
microsoft,369


<a id="inversedocumentfrequency"></a>

### Inverse document frequency (IDF)

**IDF** penalises terms that appear in many documents and rewards rare, informative terms:

$$idf_t = \log\left(\frac{N+1}{df_t+1}\right) + 1$$

- Common terms (high DF) → low IDF → less weight
- Rare terms (low DF) → high IDF → more weight

In [85]:
import numpy as np

# Total number of documents (N)
N = len(df)

# Calculate IDF
# idf = log((N+1) / (df+1)) + 1
idf = np.log((N + 1) / (document_frequency + 1)) + 1

# Create a DataFrame
df_idf_results = pd.DataFrame(idf, columns=['IDF_Score'])

# Show the results: Rare words vs. Common words

print("IDF Scores - Top 10 Highest :")
display(df_idf_results.sort_values(by='IDF_Score', ascending=False).head(10))

print("\nIDF Scores - Top 10 Lowest :")
display(df_idf_results.sort_values(by='IDF_Score', ascending=True).head(10))

IDF Scores - Top 10 Highest :


,IDF_Score
zyfwp,7.654153
0540,7.654153
yii,7.654153
yaml,7.654153
xwiki,7.654153
xvm,7.654153
xstreamhandler,7.654153
xslt,7.654153
xscreensaver,7.654153
xrwhen,7.654153



IDF Scores - Top 10 Lowest :


,IDF_Score
vulnerability,1.149116
contains,1.559455
attacker,1.624670
allows,1.717936
remote,1.793366
code,1.832587
could,2.276562
allow,2.321434
execution,2.331143
microsoft,2.433797


<a id="tfidfweighting"></a>

### TF-IDF weighting

**TF-IDF** combines both steps:

$$\text{TF-IDF}(t, d) = tf(t, d) \times idf_t$$

A term scores high only when it is **frequent in a specific document** AND **rare across the corpus** — exactly what makes it a good descriptor of that document's topic.

In [86]:
# Calculate the final TF-IDF scores
# Log-TF X IDF
# axis=1 代表將 IDF 向量的每個值對應到矩陣的「欄位 (Words)」進行計算
df_tfidf = df_log_tf.multiply(idf, axis=1)

# Show the result
sample_cve = df_tfidf.iloc[0]
print(f"Top keywords for CVE {df_tfidf.index[0]}:")
print(sample_cve.sort_values(ascending=False).head(10))

# 3. Show a small part of the final matrix
print("\nFinal TF-IDF Matrix (Sample):")
display(df_tfidf.iloc[:5, 0:10])

Top keywords for CVE CVE-2025-32432:
craft        6.401390
cm           5.862393
injection    3.272126
arbitrary    2.922350
execute      2.500861
code         2.384251
remote       1.793366
allows       1.717936
attacker     1.624670
contains     1.559455
Name: CVE-2025-32432, dtype: float64

Final TF-IDF Matrix (Sample):


,0540,10,100,100050,10103,10530,11,11014,11302,11306
cveID,,,,,,,,,,
CVE-2025-32432,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
CVE-2025-54068,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
CVE-2025-43510,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
CVE-2025-43520,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
CVE-2025-31277,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0


In [87]:
display(df_tfidf.iloc[:5, 100:110])

,acroformdll,acronis,across,action,actionform,actionmessage,actionpacklibabstractcontrollerbaserb,actionscript,actionsetup,activation
cveID,,,,,,,,,,
CVE-2025-32432,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
CVE-2025-54068,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
CVE-2025-43510,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
CVE-2025-43520,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
CVE-2025-31277,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0


<a id="cosinesimilarityretrieval"></a>

### Cosine similarity retrieval

**Cosine similarity** measures the angle between two vectors in the TF-IDF space:

$$\cos(\vec{q}, \vec{d}) = \frac{\vec{q} \cdot \vec{d}}{\|\vec{q}\| \cdot \|\vec{d}\|}$$

- Score of **1.0** = identical direction (very similar)
- Score of **0.0** = orthogonal (no shared vocabulary)

In [88]:
from sklearn.metrics.pairwise import cosine_similarity

# query: user input
# tfidf_matrix: df_tfidf
# vectorizer:  CountVectorizer
# idf_vector:  idf Score
# top_n : show top n data

def search_vulnerability(query, tfidf_matrix, vectorizer, idf_vector, top_n=5):

    # Preprocess the query
    processed_query = final_preprocess(query)
    
    # Transform query to TF vector
    query_tf = vectorizer.transform([processed_query]).toarray()
    
    # Apply Log-TF and Multiply by IDF to get Query TF-IDF
    query_log_tf = np.where(query_tf > 0, 1 + np.log10(query_tf, where=query_tf>0), 0)
    # Log-TF X IDF 
    query_tfidf = query_log_tf * idf_vector.values
    
    # Calculate Cosine Similarity between query and all documents
    # cosine_similarity feeback is  array of arrays，so use [0]
    similarities = cosine_similarity(query_tfidf, tfidf_matrix)[0]
    
    # Get the Top N results
    results_idx = similarities.argsort()[-top_n:][::-1]
    
    # creat DataFrame
    results = []
    for idx in results_idx:
        results.append({
            "CVE_ID": tfidf_matrix.index[idx],
            "Score": similarities[idx],
            "Description": df.iloc[idx]['shortDescription']
        })
    
    return pd.DataFrame(results)

# --- Test ---
user_query_1 = "Apple iOS"
search_results = search_vulnerability(user_query_1, df_tfidf, tf_vectorizer, idf)

print(f"Search Results for: '{user_query_1}'")
display(search_results)

Search Results for: 'Apple iOS'


C:\Users\ThomasLiu-PC\AppData\Local\Temp\ipykernel_34880\117305338.py:18: UserWarning: 'where' used without 'out', expect unitialized memory in output. If this is intentional, use out=None.
  query_log_tf = np.where(query_tf > 0, 1 + np.log10(query_tf, where=query_tf>0), 0)


,CVE_ID,Score,Description
0,CVE-2019-7287,0.532528,Apple iOS contains a memory corruption vulnerability which could allow an attacker to perform remote code execution.
1,CVE-2023-42824,0.526159,Apple iOS and iPadOS contain an unspecified vulnerability that allows for local privilege escalation.
2,CVE-2023-41992,0.444188,"Apple iOS, iPadOS, macOS, and watchOS contain an unspecified vulnerability that allows for local privilege escalation."
3,CVE-2025-43300,0.441656,"Apple iOS, iPadOS, and macOS contain an out-of-bounds write vulnerability in the Image I/O framework."
4,CVE-2018-4344,0.440788,"Apple iOS, macOS, tvOS, and watchOS contain a memory corruption vulnerability which can allow for code execution."


In [89]:
user_query_2 = "Banana Android"
search_results = search_vulnerability(user_query_2, df_tfidf, tf_vectorizer, idf)

print(f"Search Results for: '{user_query_2}'")
display(search_results)

Search Results for: 'Banana Android'


C:\Users\ThomasLiu-PC\AppData\Local\Temp\ipykernel_34880\117305338.py:18: UserWarning: 'where' used without 'out', expect unitialized memory in output. If this is intentional, use out=None.
  query_log_tf = np.where(query_tf > 0, 1 + np.log10(query_tf, where=query_tf>0), 0)


,CVE_ID,Score,Description
0,CVE-2021-1048,0.574773,Android kernel contains a use-after-free vulnerability that allows for privilege escalation.
1,CVE-2023-35674,0.556708,Android Framework contains an unspecified vulnerability that allows for privilege escalation.
2,CVE-2024-43093,0.556708,Android Framework contains an unspecified vulnerability that allows for privilege escalation.
3,CVE-2025-48572,0.556708,Android Framework contains an unspecified vulnerability that allows for privilege escalation.
4,CVE-2025-48633,0.507077,Android Framework contains an unspecified vulnerability that allows for information disclosure.


### Summary

* Tokenizer、Normalization、Stop-word removal、Stemming or lemmatization : Clean the messy text.
* Term-document incidence matrix、Term frequency、Log-frequency weighting、Document frequency、Inverse document frequency、TF-IDF weighting : Give every word a "value score" (TF-IDF)
* Cosine similarity retrieval : Find the best match for the user's question.


<a id="part_c"></a>
## Part C — Querying

<a id="5_informationneeds"></a>

### 5 Information Needs

| No. | Information Need | Search Query | Testing Goal |
|:---|:---|:---|:---|
| **1** | I want to find problems with Apple iPhones that let hackers control the phone from far away. | "Apple iPhone remote control" | Check if it finds Apple correctly. |
| **2** | I want to find news about Ransomware (bad software) that locks or hides files. | "Ransomware lock files" | Check if it finds bad software topics. |
| **3** | I want to find bugs in Microsoft Excel when using tables or numbers. | "Microsoft Excel table" | Check if it finds the right software. |
| **4** | I want to find routers (internet boxes) that have easy or "hidden" passwords. | "Router easy password" | Check if it finds hardware problems. |
| **5** | I want to find why Google Chrome browser sometimes crashes or has memory errors. | "Chrome browser memory error" | Check if it finds browser problems. |

In [90]:
# 5 information needs
test_query_1 = "Apple iPhone remote control"
test_query_2 = "Ransomware lock files"
test_query_3 = "Microsoft Excel table"
test_query_4 = "Router easy password"
test_query_5 = "Chrome browser memory error"


### Comparison Define Functions

In [91]:
def comparison(query, top_n=5):
    modes = ['binary', 'raw_tf', 'tfidf']
    all_results = {}
    
    # Preprocess Query
    processed_query = final_preprocess(query)
    query_tf = tf_vectorizer.transform([processed_query]).toarray()
    
    for mode in modes:
        if mode == 'binary':
            # Binary: 0 / 1
            q_vec = np.where(query_tf > 0, 1, 0)
            matrix = df_incidence.values
        elif mode == 'raw_tf':
            # Raw TF: Use times
            q_vec = query_tf
            matrix = df_tf.values
        else:
            # TF-IDF: Log-TF * IDF
            query_log_tf = np.where(query_tf > 0, 1 + np.log10(query_tf, where=query_tf>0), 0)
            q_vec = query_log_tf * idf.values
            matrix = df_tfidf.values

        # Cosine Similarity
        sims = cosine_similarity(q_vec, matrix)[0]
        idx = sims.argsort()[-top_n:][::-1]
        
        res = []
        for i in idx:
            res.append({
                "Rank": len(res) + 1,
                "CVE_ID": df.iloc[i]['cveID'],
                "Score": round(sims[i], 4),
                "Description": df.iloc[i]['shortDescription']
            })
        all_results[mode] = pd.DataFrame(res)
        
    return all_results


<a id="query_1"></a>

### Query 1 Results

| No. | Information Need | Search Query | Testing Goal |
|:---|:---|:---|:---|
| **1** | I want to find problems with Apple iPhones that let hackers control the phone from far away. | "Apple iPhone remote control" | Check if it finds Apple correctly. |

In [92]:
import pandas as pd
pd.set_option('display.max_colwidth', None)

results_q1 = comparison(test_query_1)

print(f"Information Need: {test_query_1}")
for mode in ['binary', 'raw_tf', 'tfidf']:
    print(f"\n--- Method: {mode.upper()} ---")
    display(results_q1[mode])

Information Need: Apple iPhone remote control

--- Method: BINARY ---


C:\Users\ThomasLiu-PC\AppData\Local\Temp\ipykernel_34880\3110686983.py:20: UserWarning: 'where' used without 'out', expect unitialized memory in output. If this is intentional, use out=None.
  query_log_tf = np.where(query_tf > 0, 1 + np.log10(query_tf, where=query_tf>0), 0)


,Rank,CVE_ID,Score,Description
0,1,CVE-2019-7287,0.2774,Apple iOS contains a memory corruption vulnerability which could allow an attacker to perform remote code execution.
1,2,CVE-2021-30883,0.2673,"Apple iOS, macOS, watchOS, and tvOS contain a memory corruption vulnerability that could allow for remote code execution."
2,3,CVE-2019-7238,0.2673,Sonatype Nexus Repository Manager before 3.15.0 has an incorrect access control vulnerability. Exploitation allows for remote code execution.
3,4,CVE-2024-27348,0.2673,Apache HugeGraph-Server contains an improper access control vulnerability that could allow a remote attacker to execute arbitrary code.
4,5,CVE-2025-68613,0.2500,n8n contains an improper control of dynamically managed code resources vulnerability in its workflow expression evaluation system that allows for remote code execution.



--- Method: RAW_TF ---


,Rank,CVE_ID,Score,Description
0,1,CVE-2012-1856,0.3062,The TabStrip ActiveX control in the Common Controls in MSCOMCTL.OCX in Microsoft Office allows remote attackers to execute arbitrary code via a crafted (1) document or (2) web page that triggers system-state corruption.
1,2,CVE-2017-0146,0.2774,The SMBv1 server in Microsoft Windows allows remote attackers to perform remote code execution.
2,3,CVE-2019-7287,0.2774,Apple iOS contains a memory corruption vulnerability which could allow an attacker to perform remote code execution.
3,4,CVE-2024-27348,0.2673,Apache HugeGraph-Server contains an improper access control vulnerability that could allow a remote attacker to execute arbitrary code.
4,5,CVE-2021-30883,0.2673,"Apple iOS, macOS, watchOS, and tvOS contain a memory corruption vulnerability that could allow for remote code execution."



--- Method: TFIDF ---


,Rank,CVE_ID,Score,Description
0,1,CVE-2019-18426,0.3258,A vulnerability in WhatsApp Desktop when paired with WhatsApp for iPhone allows cross-site scripting and local file reading.
1,2,CVE-2019-7287,0.1917,Apple iOS contains a memory corruption vulnerability which could allow an attacker to perform remote code execution.
2,3,CVE-2019-8526,0.1714,Apple macOS contains a use-after-free vulnerability that could allow for privilege escalation.
3,4,CVE-2024-27348,0.1582,Apache HugeGraph-Server contains an improper access control vulnerability that could allow a remote attacker to execute arbitrary code.
4,5,CVE-2023-42824,0.1553,Apple iOS and iPadOS contain an unspecified vulnerability that allows for local privilege escalation.


### Analysis of Query 1: "Apple iPhone remote control"

**Top Documents**(TF-IDF):

* CVE-2019-18426: This is relevant. It is about a WhatsApp bug on the iPhone.

* CVE-2019-7287: This is very relevant. It is an Apple iOS bug that allows remote code execution. This is exactly what I wanted to find.

Comparison of Representations:

* Binary: Not good. It found "Apache HugeGraph" (CVE-2024-27348) at Rank 4. It only sees that the words "control" and "remote" exist, but it doesn't know which document is more important.

* Raw TF: It is incorrect. It put a Microsoft Office bug at Rank 1. This happened because the word "control" appears two times in that description. Raw TF likes files that repeat words.

* TF-IDF: This is the most accurate. It knows "iPhone" is a special and important word, while "control" is a common word. That's why it correctly puts Apple-related results at the top.

<a id="query_2"></a>

### Query 2 Results

| No. | Information Need | Search Query | Testing Goal |
|:---|:---|:---|:---|
| **2** | I want to find news about Ransomware (bad software) that locks or hides files. | "Ransomware lock files" | Check if it finds bad software topics. |

In [93]:
import pandas as pd
pd.set_option('display.max_colwidth', None)

results_q2 = comparison(test_query_2)

print(f"Information Need: {test_query_2}")
for mode in ['binary', 'raw_tf', 'tfidf']:
    print(f"\n--- Method: {mode.upper()} ---")
    display(results_q2[mode])

Information Need: Ransomware lock files

--- Method: BINARY ---


C:\Users\ThomasLiu-PC\AppData\Local\Temp\ipykernel_34880\3110686983.py:20: UserWarning: 'where' used without 'out', expect unitialized memory in output. If this is intentional, use out=None.
  query_log_tf = np.where(query_tf > 0, 1 + np.log10(query_tf, where=query_tf>0), 0)


,Rank,CVE_ID,Score,Description
0,1,CVE-2022-27593,0.2887,Certain QNAP NAS running Photo Station with internet exposure contain an externally controlled reference to a resource vulnerability which can allow an attacker to modify system files. This vulnerability was observed being utilized in a Deadbolt ransomware campaign.
1,2,CVE-2020-13671,0.2500,Improper sanitization in the extension file names is present in Drupal core.
2,3,CVE-2021-38163,0.2500,SAP NetWeaver contains a vulnerability that allows unrestricted file upload.
3,4,CVE-2021-43798,0.2236,Grafana contains a path traversal vulnerability that could allow access to local files.
4,5,CVE-2019-18426,0.2132,A vulnerability in WhatsApp Desktop when paired with WhatsApp for iPhone allows cross-site scripting and local file reading.



--- Method: RAW_TF ---


,Rank,CVE_ID,Score,Description
0,1,CVE-2019-5418,0.3638,"Rails Ruby on Rails contains a path traversal vulnerability in Action View. Specially crafted accept headers in combination with calls to `render file:` can cause arbitrary files on the target server to be rendered, disclosing the file contents."
1,2,CVE-2025-11371,0.3430,Gladinet CentreStack and Triofox contains a files or directories accessible to external parties vulnerability that allows unintended disclosure of system files.
2,3,CVE-2024-57727,0.3313,SimpleHelp remote support software contains multiple path traversal vulnerabilities that allow unauthenticated remote attackers to download arbitrary files from the SimpleHelp host via crafted HTTP requests. These files may include server configuration files and hashed user passwords.
3,4,CVE-2016-3715,0.3244,"ImageMagick contains an unspecified vulnerability that could allow users to delete files by using ImageMagick's 'ephemeral' pseudo protocol, which deletes files after reading."
4,5,CVE-2021-3493,0.3244,"The overlayfs stacking file system in Linux kernel does not properly validate the application of file capabilities against user namespaces, which could lead to privilege escalation."



--- Method: TFIDF ---


,Rank,CVE_ID,Score,Description
0,1,CVE-2022-27593,0.2930,Certain QNAP NAS running Photo Station with internet exposure contain an externally controlled reference to a resource vulnerability which can allow an attacker to modify system files. This vulnerability was observed being utilized in a Deadbolt ransomware campaign.
1,2,CVE-2024-38189,0.1036,Microsoft Project contains an unspecified vulnerability that allows for remote code execution via a malicious file.
2,3,CVE-2018-15961,0.0948,Adobe ColdFusion contains an unrestricted file upload vulnerability that could allow for code execution.
3,4,CVE-2021-38163,0.0935,SAP NetWeaver contains a vulnerability that allows unrestricted file upload.
4,5,CVE-2021-43798,0.0931,Grafana contains a path traversal vulnerability that could allow access to local files.


### Analysis of Query 2: "Ransomware lock files"
**Top Documents** (TF-IDF):

* CVE-2022-27593: This is very relevant. The description clearly says it was used in a "Deadbolt ransomware campaign" to modify system files. This is exactly what I am looking for.

* CVE-2024-38189: This is partly relevant. It mentions a "malicious file" that allows hackers to control a computer. However, it does not specifically mention the word "Ransomware."

Comparison of Representations:

* Binary: It is okay, but not perfect. It found the Ransomware file at Rank 1 because the words "ransomware" and "files" are both there. But the other results (Rank 2-5) are just about "file upload" or "file reading," which have nothing to do with Ransomware.

* Raw TF: It is incorrect. It put CVE-2019-5418 at Rank 1. I checked the text, and the word "file" or "files" appears four times. Raw TF thinks this document is important just because it repeats the word "file" many times, even though it is about "disclosing contents," not Ransomware.

* TF-IDF: This is the most accurate. It gives a very high weight to the word "Ransomware" because it is a rare and special word in the database. Even though "file" is a common word, TF-IDF knows that "Ransomware" is the real key to finding the right answer.

<a id="query_3"></a>

### Query 3 Results

| No. | Information Need | Search Query | Testing Goal |
|:---|:---|:---|:---|
| **3** | I want to find bugs in Microsoft Excel when using tables or numbers. | "Microsoft Excel table" | Check if it finds the right software. |

In [94]:
import pandas as pd
pd.set_option('display.max_colwidth', None)

results_q3 = comparison(test_query_3)

print(f"Information Need: {test_query_3}")
for mode in ['binary', 'raw_tf', 'tfidf']:
    print(f"\n--- Method: {mode.upper()} ---")
    display(results_q3[mode])

Information Need: Microsoft Excel table

--- Method: BINARY ---


C:\Users\ThomasLiu-PC\AppData\Local\Temp\ipykernel_34880\3110686983.py:20: UserWarning: 'where' used without 'out', expect unitialized memory in output. If this is intentional, use out=None.
  query_log_tf = np.where(query_tf > 0, 1 + np.log10(query_tf, where=query_tf>0), 0)


,Rank,CVE_ID,Score,Description
0,1,CVE-2019-1297,0.3203,A remote code execution vulnerability exists in Microsoft Excel when the software fails to properly handle objects in memory.
1,2,CVE-2021-42292,0.3086,A security feature bypass vulnerability in Microsoft Excel would allow a local user to perform arbitrary code execution.
2,3,CVE-2009-0557,0.2801,Microsoft Office contains an object record corruption vulnerability that allows remote attackers to execute code via a crafted Excel file with a malformed record object.
3,4,CVE-2009-3129,0.2520,Microsoft Office Excel allows remote attackers to execute arbitrary code via a spreadsheet with a FEATHEADER record containing an invalid cbHdrData size element that affects a pointer offset.
4,5,CVE-2007-0671,0.2222,"Microsoft Office Excel contains a remote code execution vulnerability that can be exploited when a specially crafted Excel file is opened. This malicious file could be delivered as an email attachment or hosted on a malicious website. An attacker could leverage this vulnerability by creating a specially crafted Excel file, which, when opened, allowing an attacker to execute remote code on the affected system."



--- Method: RAW_TF ---


,Rank,CVE_ID,Score,Description
0,1,CVE-2019-1297,0.3203,A remote code execution vulnerability exists in Microsoft Excel when the software fails to properly handle objects in memory.
1,2,CVE-2021-42292,0.3086,A security feature bypass vulnerability in Microsoft Excel would allow a local user to perform arbitrary code execution.
2,3,CVE-2006-2492,0.2981,Microsoft Word and Microsoft Works Suites contain a malformed object pointer which allows attackers to execute code.
3,4,CVE-2007-0671,0.2760,"Microsoft Office Excel contains a remote code execution vulnerability that can be exploited when a specially crafted Excel file is opened. This malicious file could be delivered as an email attachment or hosted on a malicious website. An attacker could leverage this vulnerability by creating a specially crafted Excel file, which, when opened, allowing an attacker to execute remote code on the affected system."
4,5,CVE-2017-8540,0.2747,"The Microsoft Malware Protection Engine running on Microsoft Forefront and Microsoft Defender on Microsoft Windows Server 2008 SP2 and R2 SP1, Windows 7 SP1, Windows 8.1, Windows Server 2012 Gold and R2, Windows RT 8.1, Windows 10 Gold, 1511, 1607, and 1703, and Windows Server 2016, Microsoft Exchange Server 2013 and 2016, does not properly scan a specially crafted file leading to memory corruption. aka ""Microsoft Malware Protection Engine Remote Code Execution Vulnerability""."



--- Method: TFIDF ---


,Rank,CVE_ID,Score,Description
0,1,CVE-2019-1297,0.3322,A remote code execution vulnerability exists in Microsoft Excel when the software fails to properly handle objects in memory.
1,2,CVE-2021-42292,0.3316,A security feature bypass vulnerability in Microsoft Excel would allow a local user to perform arbitrary code execution.
2,3,CVE-2009-0557,0.2678,Microsoft Office contains an object record corruption vulnerability that allows remote attackers to execute code via a crafted Excel file with a malformed record object.
3,4,CVE-2007-0671,0.2378,"Microsoft Office Excel contains a remote code execution vulnerability that can be exploited when a specially crafted Excel file is opened. This malicious file could be delivered as an email attachment or hosted on a malicious website. An attacker could leverage this vulnerability by creating a specially crafted Excel file, which, when opened, allowing an attacker to execute remote code on the affected system."
4,5,CVE-2017-12319,0.1904,"A vulnerability in the Border Gateway Protocol (BGP) over an Ethernet Virtual Private Network (EVPN) for Cisco IOS XE Software could allow an unauthenticated, remote attacker to cause the device to reload, resulting in a denial of service (DoS) condition, or potentially corrupt the BGP routing table, which could result in network instability."


### Analysis of Query 3: "Microsoft Excel table"
**Top Documents** (TF-IDF):

* CVE-2019-1297: This is very relevant. It is a remote code execution vulnerability that happens exactly in Microsoft Excel. It matches my information need perfectly.

* CVE-2021-42292: This is also very relevant. It is a security bug in Microsoft Excel for local users.

Comparison of Representations:

* Binary: It is okay, but the scores are very close. It found relevant Excel files, but it also shows "Sonatype Nexus" and "Apache HugeGraph" in the top results for other queries (as we saw before) because it only checks if the word exists. It doesn't understand that "Excel" is the most important word here.

* Raw TF: It is not accurate. Look at Rank 3 (CVE-2006-2492) and Rank 5 (CVE-2017-8540). Rank 3 is about Microsoft Word, and Rank 5 is about Malware Protection Engine. They appear because they are long documents that repeat the word "Microsoft" or "file" many times. Raw TF gets "distracted" by long texts and common words.

* TF-IDF: This is the most accurate. It correctly put four Microsoft Excel vulnerabilities at the very top (Rank 1 to 4). Even though Rank 5 is about a "routing table" in Cisco, its score (0.1904) is much lower than the Excel results. TF-IDF successfully prioritized the unique word "Excel" over the common word "table."

<a id="query_4"></a>

### Query 4 Results

| No. | Information Need | Search Query | Testing Goal |
|:---|:---|:---|:---|
| **4** | I want to find routers (internet boxes) that have easy or "hidden" passwords. | "Router easy password" | Check if it finds hardware problems. |

In [95]:
import pandas as pd
pd.set_option('display.max_colwidth', None)

results_q4 = comparison(test_query_4)

print(f"Information Need: {test_query_4}")
for mode in ['binary', 'raw_tf', 'tfidf']:
    print(f"\n--- Method: {mode.upper()} ---")
    display(results_q4[mode])

Information Need: Router easy password

--- Method: BINARY ---


C:\Users\ThomasLiu-PC\AppData\Local\Temp\ipykernel_34880\3110686983.py:20: UserWarning: 'where' used without 'out', expect unitialized memory in output. If this is intentional, use out=None.
  query_log_tf = np.where(query_tf > 0, 1 + np.log10(query_tf, where=query_tf>0), 0)


,Rank,CVE_ID,Score,Description
0,1,CVE-2011-4723,0.4082,"The D-Link DIR-300 router stores cleartext passwords, which allows context-dependent attackers to obtain sensitive information."
1,2,CVE-2021-40655,0.3430,D-Link DIR-605 routers contain an information disclosure vulnerability that allows attackers to obtain a username and password by forging a post request to the /getcfg.php page.
2,3,CVE-2018-6530,0.2357,Multiple D-Link routers contain an unspecified vulnerability that allows for execution of OS commands.
3,4,CVE-2017-6077,0.2236,NETGEAR DGN2200 wireless routers contain a vulnerability that allows for remote code execution.
4,5,CVE-2016-10174,0.2132,The NETGEAR WNR2000v5 router contains a buffer overflow which can be exploited to achieve remote code execution.



--- Method: RAW_TF ---


,Rank,CVE_ID,Score,Description
0,1,CVE-2011-4723,0.4082,"The D-Link DIR-300 router stores cleartext passwords, which allows context-dependent attackers to obtain sensitive information."
1,2,CVE-2021-40655,0.3430,D-Link DIR-605 routers contain an information disclosure vulnerability that allows attackers to obtain a username and password by forging a post request to the /getcfg.php page.
2,3,CVE-2014-100005,0.3162,D-Link DIR-600 routers contain a cross-site request forgery (CSRF) vulnerability that allows an attacker to change router configurations by hijacking an existing administrator session.
3,4,CVE-2019-1653,0.2722,Cisco Small Business RV320 and RV325 Dual Gigabit WAN VPN Routers contain improper access controls for URLs. Exploitation could allow an attacker to download the router configuration or detailed diagnostic information.
4,5,CVE-2026-23760,0.2652,"SmarterTools SmarterMail contains an authentication bypass using an alternate path or channel vulnerability in the password reset API. The force-reset-password endpoint permits anonymous requests and fails to verify the existing password or a reset token when resetting system administrator accounts. This could allow an unauthenticated attacker to supply a target administrator username and a new password to reset the account, resulting in full administrative compromise of the SmarterMail instance."



--- Method: TFIDF ---


,Rank,CVE_ID,Score,Description
0,1,CVE-2011-4723,0.3618,"The D-Link DIR-300 router stores cleartext passwords, which allows context-dependent attackers to obtain sensitive information."
1,2,CVE-2021-40655,0.3389,D-Link DIR-605 routers contain an information disclosure vulnerability that allows attackers to obtain a username and password by forging a post request to the /getcfg.php page.
2,3,CVE-2018-6530,0.3294,Multiple D-Link routers contain an unspecified vulnerability that allows for execution of OS commands.
3,4,CVE-2017-6077,0.2411,NETGEAR DGN2200 wireless routers contain a vulnerability that allows for remote code execution.
4,5,CVE-2019-16920,0.2285,Multiple D-Link routers contain a command injection vulnerability which can allow attackers to achieve full system compromise.


### Analysis of Query 4: "Router easy password"
**Top Documents** (TF-IDF):

* CVE-2011-4723: This is very relevant. It says the D-Link router stores "cleartext passwords." This is a big security problem and exactly what I wanted to find.

* CVE-2021-40655: This is also very relevant. It says attackers can get the "username and password" from a D-Link router.

Comparison of Representations:

* Binary: It is okay. It successfully found the D-Link router at Rank 1 and 2. However, the scores for other routers (NETGEAR) are very low and similar. It doesn't show a big difference between a "perfect match" and a "partial match."

* Raw TF: It is incorrect at Rank 5. Look at CVE-2026-23760. This is about "SmarterMail" (an email software), not a router. It appeared in the results because the word "password" is repeated 5 times in that description (password reset, existing password, new password, etc.). Raw TF was "tricked" by the word count.

* TF-IDF: This is the most accurate. Even though "password" is a common word, TF-IDF kept the focus on "Router." All top 5 results in TF-IDF are about routers (D-Link and NETGEAR). It successfully ignored the email software (SmarterMail) that Raw TF incorrectly suggested.

<a id="query_5"></a>

### Query 5 Results

| No. | Information Need | Search Query | Testing Goal |
|:---|:---|:---|:---|
| **5** | I want to find why Google Chrome browser sometimes crashes or has memory errors. | "Chrome browser memory error" | Check if it finds browser problems. |

In [96]:
import pandas as pd
pd.set_option('display.max_colwidth', None)

results_q5 = comparison(test_query_5)

print(f"Information Need: {test_query_5}")
for mode in ['binary', 'raw_tf', 'tfidf']:
    print(f"\n--- Method: {mode.upper()} ---")
    display(results_q5[mode])

Information Need: Chrome browser memory error

--- Method: BINARY ---


C:\Users\ThomasLiu-PC\AppData\Local\Temp\ipykernel_34880\3110686983.py:20: UserWarning: 'where' used without 'out', expect unitialized memory in output. If this is intentional, use out=None.
  query_log_tf = np.where(query_tf > 0, 1 + np.log10(query_tf, where=query_tf>0), 0)


,Rank,CVE_ID,Score,Description
0,1,CVE-2024-4761,0.2942,"Google Chromium V8 Engine contains an unspecified out-of-bounds memory write vulnerability via a crafted HTML page. This vulnerability could affect multiple web browsers that utilize Chromium, including, but not limited to, Google Chrome, Microsoft Edge, and Opera."
1,2,CVE-2021-38003,0.2835,"Google Chromium V8 Engine has a bug in JSON.stringify, where the internal TheHole value can leak to script code, causing memory corruption. This vulnerability could affect multiple web browsers that utilize Chromium, including, but not limited to, Google Chrome, Microsoft Edge, and Opera."
2,3,CVE-2025-14174,0.2835,"Google Chromium contains an out of bounds memory access vulnerability in ANGLE that could allow a remote attacker to perform out of bounds memory access via a crafted HTML page. This vulnerability could affect multiple web browsers that utilize Chromium, including, but not limited to, Google Chrome, Microsoft Edge, and Opera."
3,4,CVE-2017-5030,0.2785,"Google Chromium V8 Engine contains a memory corruption vulnerability that allows a remote attacker to execute code via a crafted HTML page. This vulnerability could affect multiple web browsers that utilize Chromium, including, but not limited to, Google Chrome, Microsoft Edge, and Opera."
4,5,CVE-2025-2783,0.2785,"Google Chromium Mojo on Windows contains a sandbox escape vulnerability caused by a logic error, which results from an incorrect handle being provided in unspecified circumstances. This vulnerability could affect multiple web browsers that utilize Chromium, including, but not limited to, Google Chrome, Microsoft Edge, and Opera."



--- Method: RAW_TF ---


,Rank,CVE_ID,Score,Description
0,1,CVE-2024-43047,0.3198,Multiple Qualcomm chipsets contain a use-after-free vulnerability due to memory corruption in DSP Services while maintaining memory maps of HLOS memory.
1,2,CVE-2021-37976,0.2887,"Google Chromium contains an information disclosure vulnerability within the core memory component that allows a remote attacker to obtain potentially sensitive information from process memory via a crafted HTML page. This vulnerability could affect multiple web browsers that utilize Chromium, including, but not limited to, Google Chrome, Microsoft Edge, and Opera."
2,3,CVE-2025-14174,0.2857,"Google Chromium contains an out of bounds memory access vulnerability in ANGLE that could allow a remote attacker to perform out of bounds memory access via a crafted HTML page. This vulnerability could affect multiple web browsers that utilize Chromium, including, but not limited to, Google Chrome, Microsoft Edge, and Opera."
3,4,CVE-2026-21385,0.2774,Multiple Qualcomm chipsets contain a memory corruption vulnerability while using alignments for memory allocation.
4,5,CVE-2021-38003,0.2572,"Google Chromium V8 Engine has a bug in JSON.stringify, where the internal TheHole value can leak to script code, causing memory corruption. This vulnerability could affect multiple web browsers that utilize Chromium, including, but not limited to, Google Chrome, Microsoft Edge, and Opera."



--- Method: TFIDF ---


,Rank,CVE_ID,Score,Description
0,1,CVE-2023-36874,0.3318,Microsoft Windows Error Reporting Service contains an unspecified vulnerability that allows for privilege escalation.
1,2,CVE-2025-2783,0.3071,"Google Chromium Mojo on Windows contains a sandbox escape vulnerability caused by a logic error, which results from an incorrect handle being provided in unspecified circumstances. This vulnerability could affect multiple web browsers that utilize Chromium, including, but not limited to, Google Chrome, Microsoft Edge, and Opera."
2,3,CVE-2017-5030,0.2695,"Google Chromium V8 Engine contains a memory corruption vulnerability that allows a remote attacker to execute code via a crafted HTML page. This vulnerability could affect multiple web browsers that utilize Chromium, including, but not limited to, Google Chrome, Microsoft Edge, and Opera."
3,4,CVE-2024-4761,0.2642,"Google Chromium V8 Engine contains an unspecified out-of-bounds memory write vulnerability via a crafted HTML page. This vulnerability could affect multiple web browsers that utilize Chromium, including, but not limited to, Google Chrome, Microsoft Edge, and Opera."
4,5,CVE-2024-26169,0.2546,Microsoft Windows Error Reporting Service contains an improper privilege management vulnerability that allows a local attacker with user permissions to gain SYSTEM privileges.


### Analysis of Query 5: "Chrome browser memory error"
**Top Documents** (TF-IDF):

* CVE-2023-36874: This is not relevant. It is about "Microsoft Windows Error Reporting Service." It appeared at Rank 1 because it matches the word "Error" very strongly, but it has nothing to do with the Chrome browser.

* CVE-2025-2783: This is very relevant. It is about "Google Chromium" (which is the base of the Chrome browser) and a security logic error.

Comparison of Representations:

* Binary: Surprisingly, Binary performed very well here. All top 5 results are about "Google Chromium." Because these documents contain all the words ("Chrome," "memory," and "error"), they all got a similar high score. In this specific case, Binary was very stable.

* Raw TF: It is incorrect at Rank 1. It found a "Qualcomm chipset" bug. This happened because the word "memory" appears many times in that description (HLOS memory, DSP services memory, etc.). Raw TF was "tricked" by the word count again.

* TF-IDF: It is mostly accurate, but it got a bit confused. It put "Microsoft Windows Error Reporting" at Rank 1 and 5. This is because "Error Reporting" is a very specific name in the database, so the system thought the word "error" was the most important part of your search. However, the results in Rank 2, 3, and 4 are all excellent Chrome-related memory bugs.

### Final Summary

1. **Binary Representation** (0 or 1)

    **How it works**: It only checks if a word exists (1) or not (0). It does not care how many times a word appears.

    **Performance**: Average.

    **Observation**: In Query 5 (Chrome), it worked surprisingly well because it found documents containing all my keywords. However, in most cases, it gives many different documents the same score. It cannot tell which document is "more important," so the ranking is not very helpful.

2. **Raw TF** (Term Frequency)

    **How it works**: It counts how many times a word appears in a document.

    **Performance**: Poor (Lowest accuracy).

    **Observation**: This method is easily "tricked" by long documents. For example, in Query 4 (Router), a long email software description appeared at Rank 1 simply because it repeated the word "password" many times. Raw TF likes wordy files, even if they are not relevant to the topic.

3. **TF-IDF** (Term Frequency-Inverse Document Frequency)

    **How it works**: It gives high scores to unique words (like "iPhone" or "Ransomware") and low scores to common words (like "control" or "files") that appear everywhere.

    **Performance**: Best (Highest accuracy).

    **Observation**: This was the smartest method. It successfully filtered out irrelevant results in Query 1, 2, and 3. It understands the "meaning" of the search better by focusing on the most important technical terms. Even when it made a small mistake (like in Query 5 with "Error Reporting"), the overall results were much more professional and accurate.

After comparing all three methods in this query, **TF-IDF** with Cosine Similarity is the best system for a CVE search engine.

<a id="part_d"></a>

## Part D — Evaluation

1. **Relevance Judgment**:
    A human looks at the search results and decides if a document is Relevant (1) or Irrelevant (0) based on the search goal.
2. **Confusion Matrix**:A table used to describe the performance of the system. It shows:
    * True Positive (TP): System found it, and it is correct.
    * False Positive (FP): System found it, but it is wrong (Spam).
    * False Negative (FN): System missed it (It was correct but not found).
    * True Negative (TN): System ignored it, and it was indeed wrong.
3. **Precision**:Out of everything the system found, how much is actually correct? (Quality)
4. **Recall**:Out of all the correct documents in the database, how many did the system find? (Quantity)
5. **F1-score**:The average  between Precision and Recall.
6. **Precision@K & @Top**:Precision calculated only for the top K results (e.g., Top 5 or Top 10).
7. **Average Precision (AP)**:This measure cares about ranking. It gives a higher score if the relevant documents are at the very top (Rank 1, 2) instead of the bottom.
8. **Mean Reciprocal Rank (MRR)**:It focuses only on the first correct answer. If the first right answer is at Rank 1, the score is $1/1 = 1$. If it is at Rank 2, the score is $1/2 = 0.5$.
9. **Cohen's Kappa Measure**:A score that shows how much two humans (User A and User B) agree on the results. If both say "this is relevant," the Kappa score is high.
10. **Smart Notation**:A 3-letter code used to describe the Term Frequency (TF) / Document Frequency (IDF) / Normalization.
    * TF: n:natural / l:logarithm: $1 + \log(tf)$
    * IDF: n:none / t:idf: $\log(N/df)$
    * Nor: n:none / c:cosine

<a id="query_1e"></a>

### Query 1

In [97]:
import numpy as np
import pandas as pd
from sklearn.metrics import confusion_matrix, precision_score, recall_score, f1_score, accuracy_score, cohen_kappa_score

# ==========================================================
# 1. Human Relevance Judgments (Ground Truth)
# 人工相關性判斷（地面實況）
# ==========================================================
# Based on our TF-IDF results for Query 1:
# Rank 1: CVE-2019-18426 (iPhone) -> Relevant (1)
# Rank 2: CVE-2019-7287 (Apple iOS) -> Relevant (1)
# Rank 3: CVE-2019-8526 (macOS) -> Partially Relevant (Assume 0 for strict evaluation)
# Rank 4: CVE-2024-27348 (Apache) -> Irrelevant (0)
# Rank 5: CVE-2023-42824 (Apple iOS) -> Relevant (1)

y_true = [1, 1, 0, 0, 1]  # Actual relevance (人工判斷的真實結果)
y_pred = [1, 1, 1, 1, 1]  # System prediction (系統檢索出來的都視為 Positive)

# ==========================================================
# 2. Basic Metrics Calculation (基礎指標計算)
# ==========================================================
# Precision: Quality of results | 精確率：結果的品質
precision = precision_score(y_true, y_pred)

# Recall: Completeness (Assume we found all relevant docs in top 5 for this lab)
# 召回率：完整性（假設資料庫中相關的都在這 5 筆內）
recall = recall_score(y_true, y_pred)

# F1-score: Harmonic mean of Precision and Recall | F1-分數：精確率與召回率的平衡
f1 = f1_score(y_true, y_pred)

# Accuracy: Overall correctness | 準確度：整體正確比例
accuracy = accuracy_score(y_true, y_pred)

# ==========================================================
# 3. IR Specific Metrics (資訊檢索專用指標)
# ==========================================================
# Precision@K (K=5) & @Top
p_at_k = sum(y_true) / len(y_true)

# Average Precision (AP): Considers the rank of relevant documents
# 平均精確率 (AP)：考慮相關文件出現的排名順序
def calculate_ap(labels):
    precisions = []
    relevant_found = 0
    for i, l in enumerate(labels):
        if l == 1:
            relevant_found += 1
            precisions.append(relevant_found / (i + 1))
    return np.mean(precisions) if precisions else 0

ap = calculate_ap(y_true)

# Mean Reciprocal Rank (MRR): 1 / Rank of the first relevant document
# 平均倒數排名 (MRR)：1 / 第一個正確答案的排名
first_relevant_rank = y_true.index(1) + 1
mrr = 1 / first_relevant_rank

# ==========================================================
# 4. Agreement Metric (一致性指標)
# ==========================================================
# Cohen's Kappa: Agreement between two annotators (e.g., you and a teammate)
# 柯恩卡帕係數：兩位標記人員（你與隊友）的一致性
user_a = [1, 1, 0, 0, 1] # Your labels
user_b = [1, 1, 1, 0, 1] # Teammate's labels (Simulation)
kappa = cohen_kappa_score(user_a, user_b)

# ==========================================================
# 5. Display Results (顯示結果)
# ==========================================================
print(f"--- Query 1: Evaluation Metrics ---")
print(f"Smart Notation: ltc (Log-TF, IDF, Cosine)")
print(f"-----------------------------------")
print(f"Confusion Matrix:\n{confusion_matrix(y_true, y_pred, labels=[1,0])}")
print(f"Precision: {precision:.4f}")
print(f"Recall: {recall:.4f}")
print(f"F1-score: {f1:.4f}")
print(f"Accuracy: {accuracy:.4f}")
print(f"Precision@5 (Top): {p_at_k:.4f}")
print(f"Average Precision (AP): {ap:.4f}")
print(f"MRR: {mrr:.4f}")
print(f"Cohen's Kappa: {kappa:.4f}")

--- Query 1: Evaluation Metrics ---
Smart Notation: ltc (Log-TF, IDF, Cosine)
-----------------------------------
Confusion Matrix:
[[3 0]
 [2 0]]
Precision: 0.6000
Recall: 1.0000
F1-score: 0.7500
Accuracy: 0.6000
Precision@5 (Top): 0.6000
Average Precision (AP): 0.8667
MRR: 1.0000
Cohen's Kappa: 0.5455


* Confusion Matrix [[3 0], [2 0]]:
    * TP (True Positive) = 3
    * FP (False Positive) = 2
    * FN (False Negative) = 0

* Precision 0.60: 60% of the top 5 results are relevant.
* Recall 1.00: The system found all relevant documents we expected in the top 5.
* Precision@5 (Top) 0.60: This matches our general precision, showing the "first page" of results is reliable
* Average Precision (AP) 0.8667:This is a very high score. It means the relevant documents are ranked near the top (Rank 1 and 2). AP rewards systems that put correct answers at the very beginning.
* MRR (Mean Reciprocal Rank) 1.00: It means the very first result (Rank 1) was a correct relevant document.
* Smart Notation (ltc): Confirms we used Log-TF (l), IDF weight (t), and Cosine Normalization (c).

For Query 1, the TF-IDF (ltc) model performed excellently. With an MRR of 1.0 and an AP of 0.8667, the system proved it can place the most relevant Apple vulnerabilities at the very top of the list, providing a great user experience.

<a id="query_2e"></a>

### Query 2

In [98]:
import numpy as np
from sklearn.metrics import confusion_matrix, precision_score, recall_score, f1_score, accuracy_score, cohen_kappa_score

# ==========================================================
# 1. Human Relevance Judgments for Query 2 (Ransomware)
# Query 2 人工相關性判斷
# ==========================================================
# Based on TF-IDF results for "Ransomware lock files":
# Rank 1: CVE-2022-27593 (Deadbolt ransomware) -> Relevant (1)
# Rank 2: CVE-2024-38189 (Malicious file/RCE) -> Partially Relevant (0 for strict)
# Rank 3: CVE-2018-15961 (File upload) -> Irrelevant (0)
# Rank 4: CVE-2021-38163 (File upload) -> Irrelevant (0)
# Rank 5: CVE-2021-43798 (Local files access) -> Irrelevant (0)

y_true_q2 = [1, 0, 0, 0, 0]  # Actual relevance (真實情況：只有第一名命中勒索軟體)
y_pred_q2 = [1, 1, 1, 1, 1]  # System results (前五名皆視為檢索結果)

# ==========================================================
# 2. Metrics Calculation (指標計算)
# ==========================================================
precision_q2 = precision_score(y_true_q2, y_pred_q2)
recall_q2 = recall_score(y_true_q2, y_pred_q2)
f1_q2 = f1_score(y_true_q2, y_pred_q2)
accuracy_q2 = accuracy_score(y_true_q2, y_pred_q2)

# Precision@5 (Top)
p_at_5_q2 = sum(y_true_q2) / len(y_true_q2)

# Average Precision (AP)
def calculate_ap(labels):
    precisions = []
    relevant_found = 0
    for i, l in enumerate(labels):
        if l == 1:
            relevant_found += 1
            precisions.append(relevant_found / (i + 1))
    return np.mean(precisions) if precisions else 0
ap_q2 = calculate_ap(y_true_q2)

# MRR (Mean Reciprocal Rank)
mrr_q2 = 1 / (y_true_q2.index(1) + 1)

# Cohen's Kappa (Agreement Simulation)
# 模擬兩個人判斷，假設第二個人覺得 Rank 2 也算相關
user_a_q2 = [1, 0, 0, 0, 0] 
user_b_q2 = [1, 1, 0, 0, 0] 
kappa_q2 = cohen_kappa_score(user_a_q2, user_b_q2)

# ==========================================================
# 3. Print Results (輸出結果)
# ==========================================================
print(f"--- Query 2: Evaluation Metrics (Ransomware) ---")
print(f"Smart Notation: ltc (Log-TF, IDF, Cosine)")
print(f"-----------------------------------")
print(f"Confusion Matrix:\n{confusion_matrix(y_true_q2, y_pred_q2, labels=[1,0])}")
print(f"Precision: {precision_q2:.4f}")
print(f"Recall: {recall_q2:.4f}")
print(f"F1-score: {f1_q2:.4f}")
print(f"Accuracy: {accuracy_q2:.4f}")
print(f"Precision@5 (Top): {p_at_5_q2:.4f}")
print(f"Average Precision (AP): {ap_q2:.4f}")
print(f"MRR: {mrr_q2:.4f}")
print(f"Cohen's Kappa: {kappa_q2:.4f}")

--- Query 2: Evaluation Metrics (Ransomware) ---
Smart Notation: ltc (Log-TF, IDF, Cosine)
-----------------------------------
Confusion Matrix:
[[1 0]
 [4 0]]
Precision: 0.2000
Recall: 1.0000
F1-score: 0.3333
Accuracy: 0.2000
Precision@5 (Top): 0.2000
Average Precision (AP): 1.0000
MRR: 1.0000
Cohen's Kappa: 0.5455


* Confusion Matrix [[1 0], [4 0]]:
    * TP (True Positive) = 1
    * FP (False Positive) = 4

* Precision 0.20: Only 20% of the results are relevant due to the common word "files."
* Recall 1.00: The specific ransomware document in the database was found.
* Precision@5 (Top) 0.20: Although the top 5 contain irrelevant files, the ranking algorithm correctly prioritized the rare term "Ransomware" over the common term "files."
* Average Precision (AP) 1.00:Since the only relevant document was at Rank 1, the ranking quality is perfect.
* MRR (Mean Reciprocal Rank) 1.00: : The system successfully prioritized the unique term "Ransomware" at the very top.
* Smart Notation (ltc): Confirms we used Log-TF (l), IDF weight (t), and Cosine Normalization (c).

For Query 2, while the Precision was lower due to common keywords like 'files', the system achieved a perfect MRR and AP of 1.0. This demonstrates that TF-IDF is exceptionally good at ranking the most unique and critical information at the top, even in a noisy dataset.

<a id="query_3e"></a>

### Query 3

In [99]:
import numpy as np
from sklearn.metrics import confusion_matrix, precision_score, recall_score, f1_score, accuracy_score, cohen_kappa_score

# ==========================================================
# 1. Human Relevance Judgments for Query 3 (Excel)
# Query 3 人工相關性判斷
# ==========================================================
# Based on TF-IDF results for "Microsoft Excel table":
# Rank 1: CVE-2019-1297 (Excel RCE) -> Relevant (1)
# Rank 2: CVE-2021-42292 (Excel Bypass) -> Relevant (1)
# Rank 3: CVE-2009-0557 (Office Excel) -> Relevant (1)
# Rank 4: CVE-2007-0671 (Office Excel) -> Relevant (1)
# Rank 5: CVE-2017-12319 (Cisco BGP Table) -> Irrelevant (0)

y_true_q3 = [1, 1, 1, 1, 0]  # Actual relevance (前四名皆為 Excel 相關)
y_pred_q3 = [1, 1, 1, 1, 1]  # System results (前五名皆視為檢索結果)

# ==========================================================
# 2. Metrics Calculation (指標計算)
# ==========================================================
precision_q3 = precision_score(y_true_q3, y_pred_q3)
recall_q3 = recall_score(y_true_q3, y_pred_q3)
f1_q3 = f1_score(y_true_q3, y_pred_q3)
accuracy_q3 = accuracy_score(y_true_q3, y_pred_q3)

# Precision@5 (Top)
p_at_5_q3 = sum(y_true_q3) / len(y_true_q3)

# Average Precision (AP)
def calculate_ap(labels):
    precisions = []
    relevant_found = 0
    for i, l in enumerate(labels):
        if l == 1:
            relevant_found += 1
            precisions.append(relevant_found / (i + 1))
    return np.mean(precisions) if precisions else 0
ap_q3 = calculate_ap(y_true_q3)

# MRR (Mean Reciprocal Rank)
mrr_q3 = 1 / (y_true_q3.index(1) + 1)

# Cohen's Kappa (Agreement Simulation)
# 模擬兩個人判斷，假設第二個人對 Rank 5 的 "Table" 定義不同
user_a_q3 = [1, 1, 1, 1, 0] 
user_b_q3 = [1, 1, 1, 1, 1] 
kappa_q3 = cohen_kappa_score(user_a_q3, user_b_q3)

# ==========================================================
# 3. Print Results (輸出結果)
# ==========================================================
print(f"--- Query 3: Evaluation Metrics (Excel) ---")
print(f"Smart Notation: ltc (Log-TF, IDF, Cosine)")
print(f"-----------------------------------")
print(f"Confusion Matrix:\n{confusion_matrix(y_true_q3, y_pred_q3, labels=[1,0])}")
print(f"Precision: {precision_q3:.4f}")
print(f"Recall: {recall_q3:.4f}")
print(f"F1-score: {f1_q3:.4f}")
print(f"Accuracy: {accuracy_q3:.4f}")
print(f"Precision@5 (Top): {p_at_5_q3:.4f}")
print(f"Average Precision (AP): {ap_q3:.4f}")
print(f"MRR: {mrr_q3:.4f}")
print(f"Cohen's Kappa: {kappa_q3:.4f}")

--- Query 3: Evaluation Metrics (Excel) ---
Smart Notation: ltc (Log-TF, IDF, Cosine)
-----------------------------------
Confusion Matrix:
[[4 0]
 [1 0]]
Precision: 0.8000
Recall: 1.0000
F1-score: 0.8889
Accuracy: 0.8000
Precision@5 (Top): 0.8000
Average Precision (AP): 1.0000
MRR: 1.0000
Cohen's Kappa: 0.0000


* Confusion Matrix [[1 0], [4 0]]:
    * TP (True Positive) = 4
    * FP (False Positive) = 1
    * FN (False Negative) = 0
* Precision 0.80: 80% of the results are highly relevant.
* Recall 1.00: All relevant documents for this specific software were retrieved.
* Precision@5 (Top) 0.80: Demonstrates high accuracy for combined keyword queries.
* Average Precision (AP) 1.00:All relevant documents were clustered perfectly at the top (Ranks 1-4)
* MRR (Mean Reciprocal Rank) 1.00: The first result was a perfect match for Microsoft Excel.
* Smart Notation (ltc): Confirms we used Log-TF (l), IDF weight (t), and Cosine Normalization (c).

For Query 3, the system showed high precision and perfect ranking. This proves that the TF-IDF (ltc) model is very effective at handling queries with multiple specific keywords, ensuring that relevant software-specific bugs are prioritized.

### Reflection
Answer these questions:

1. Which representation worked best and why?

    TF-IDF (ltc) worked best. Unlike Binary or Raw TF, it can distinguish between "common words" and "critical keywords." In our tests, it successfully prioritized unique security terms like "Ransomware" or "iPhone" while lowering the weight of common words like "file" or "system."

2. Did TF-IDF improve over raw term counts?

    Yes, significantly. Raw TF often ranked long documents higher just because they repeated certain words many times (e.g., Query 4). TF-IDF fixed this by using Inverse Document Frequency (IDF) to penalize words that appear in too many documents, ensuring that specific vulnerabilities are ranked at the top.

3. What kinds of false positives did you observe?

    Most false positives came from "Polysemy" (words with multiple meanings). For example, in Query 3 ("Excel table"), the system retrieved a Cisco "routing table." Although the word "table" matched, the context (networking vs. spreadsheet) was completely different.

4. What kinds of relevant documents were missed?

    We missed documents that used synonyms instead of exact keywords. For instance, a query for "memory error" might miss documents that only use technical terms like "buffer overflow" or "heap corruption" without explicitly saying "memory error."

5. How did the evaluation metrics help you understand system quality?

    Metrics like MRR and AP were much more useful than simple Accuracy. They showed us that our system is excellent at ranking the best result at Rank 1, even if the overall Precision is low. This helps us understand the "User Experience" (finding the answer quickly) rather than just counting correct words.

6. How would you improve the system next?

    I would implement Stemming or Lemmatization to handle different word forms (e.g., "locking" vs "locked"). I would also add a Thesaurus/Synonym mapping to connect general terms like "memory error" to specific technical terms like "segmentation fault." Finally, using N-grams would help capture phrases like "remote code execution" as a single unit.
